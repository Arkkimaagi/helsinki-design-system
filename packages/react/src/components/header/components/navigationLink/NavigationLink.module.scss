@use "../../Header.module.scss" as header;
@import "../../../../styles/common.scss";

.navigationLink.navigationLink {
  --link-color: var(--color-black-90);
  --link-font-size: 14px;
  --link-visited-color: none;

  display: flex;
  flex-grow: 1;
  line-height: var(--lineheight-l);
  padding: 13px 0 15px;
  margin: 0 calc(-1 * var(--header-focus-outline-width));
  text-decoration: none;
  word-wrap: break-word;

  &:hover,
  &:focus {
    text-decoration: underline;
    text-decoration-color: var(--color-black-90);
  }
}

.active {
  font-weight: 700;
  padding-left: var(--header-focus-outline-width);
}


.navigationLinkWrapper {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: stretch;
  position: relative;

  &.isNotLargeScreen {
    align-items: stretch;
    border-top: 1px solid var(--color-black-20);

    &,
    &:focus {
      border-top-width: 1px;
    }

    ul {
      border: none;
    }
  }
}


.navigationLink.isNotLargeScreen{
  align-self: center;
  border: none;
  box-sizing: border-box;
  color: var(--color-black-90);
  font-size: var(--link-font-size);
  height: var(--action-bar-dropdown-menu-secondary-item-height);
  line-height: var(--action-bar-dropdown-menu-secondary-item-height);
  padding: 0;
  word-wrap: break-word;

  &.depth-0 {
    font-size: 20px;
    height: var(--action-bar-dropdown-menu-primary-item-height);
    line-height: var(--action-bar-dropdown-menu-primary-item-height);
    padding: 0 0 0 16px;
  }

  &.depth-1 { padding: 0 0 0 32px; }
  &.depth-2 { padding: 0 0 0 48px; }
  &.depth-3 { padding: 0 0 0 64px; }

  &:focus { border: none }

  @include header.element-focus-state;
}
