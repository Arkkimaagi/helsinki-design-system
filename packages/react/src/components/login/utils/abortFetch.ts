export function isAbortError(error: Error): boolean {
  return !!error && error.name === 'AbortError';
}

export function createFetchAborter() {
  let isAborted = false;
  let abortController: AbortController | undefined;
  const abort = () => {
    if (isAborted || !abortController) {
      return;
    }
    isAborted = true;
    abortController.abort();
    abortController = undefined;
  };
  return {
    isAborted: () => isAborted,
    getSignal: () => {
      abort();
      isAborted = false;
      abortController = new AbortController();
      return abortController.signal;
    },
    abort,
  };
}
