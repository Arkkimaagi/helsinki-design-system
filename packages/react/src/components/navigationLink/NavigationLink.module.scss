.navigationLink.navigationLink  {
    --link-visited-color: none;
    --link-color: var(--color-black-90);
    text-decoration: none;
}
