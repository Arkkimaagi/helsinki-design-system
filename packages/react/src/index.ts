// import core base styles
// import 'hds-core';

export * from './icons';
export * from './components';
