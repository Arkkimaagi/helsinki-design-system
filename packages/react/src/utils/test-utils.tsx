import React, { PropsWithChildren } from 'react';

import { Navigation } from '../components/navigation';
import { Footer } from '../components/footer';

export const NavigationWrapper = ({ children }: PropsWithChildren<{}>) => (
  <Navigation
    menuCloseAriaLabel="close"
    menuOpenAriaLabel="open"
    skipTo="#content"
    skipToContentLabel="Skip to content"
    title="Foo"
  >
    {children}
  </Navigation>
);

export const FooterWrapper = ({ children }: PropsWithChildren<{}>) => (
  <Footer footerAriaLabel="Foo" title="Bar">
    {children}
  </Footer>
);
