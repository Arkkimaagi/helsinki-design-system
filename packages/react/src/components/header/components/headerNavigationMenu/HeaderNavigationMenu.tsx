import React, { Children, cloneElement, isValidElement, useEffect } from 'react';

// import base styles
import '../../../../styles/base.css';

import { useHeaderContext, useSetHeaderContext } from '../../HeaderContext';
import classNames from '../../../../utils/classNames';
import { getChildElementsEvenIfContainersInbetween } from '../../../../utils/getChildren';
import { HeaderNavigationMenuContextProvider } from './HeaderNavigationMenuContext';
import styles from './HeaderNavigationMenu.module.scss';

export type HeaderNavigationMenuProps = React.PropsWithChildren<{
  /**
   * aria-label for describing universal bar.
   */
  ariaLabel?: string;
  /**
   * Children are expected to be NavigationLink components or a container with NavigationLink components inside.
   */
  children?: React.ReactNode;
  /**
   * ID of the header element.
   */
  id?: string;
}>;

const renderHeaderNavigationMenuItem = (child, index) => {
  const { isNotLargeScreen } = useHeaderContext();
  const linkContentClass = isNotLargeScreen
    ? styles.headerNavigationMenuLinkContentMobile
    : styles.headerNavigationMenuLinkContent;

  const activeLinkClassName = classNames(linkContentClass, styles.headerNavigationMenuLinkContentActive);
  const linkContainerClasses = child.props.active && !isNotLargeScreen ? activeLinkClassName : linkContentClass;

  // Pass several className props downwards
  const mobileNode = cloneElement(child as React.ReactElement, {
    dropdownLinkClassName: styles.headerNavigationMenuDropdownLinkMobile,
    dropdownClassName: styles.headerNavigationMenuDropdownMobile,
    wrapperClassName: styles.headerNavigationMenuLinkWrapperMobile,
    dropdownButtonClassName: styles.headerNavigationMenuDropdownButtonMobile,
    className: classNames(child.props.className, styles.headerNavigationMenuLinkMobile),
    index,
  });

  const desktopNode = cloneElement(child as React.ReactElement, {
    className: classNames(child.props.className, styles.headerNavigationMenuLink),
    index,
  });

  const node = isNotLargeScreen ? mobileNode : desktopNode;

  return (
    // eslint-disable-next-line react/no-array-index-key
    <li key={index}>
      <span className={linkContainerClasses}>{node}</span>
    </li>
  );
};

export const HeaderNavigationMenuContent = () => {
  const { navigationContent } = useHeaderContext();
  return (
    <>
      {Children.map(navigationContent, (child, index) => {
        if (!isValidElement(child)) return null;
        return renderHeaderNavigationMenuItem(child, index);
      })}
    </>
  );
};

export const HeaderNavigationMenu = ({ ariaLabel, children, id }: HeaderNavigationMenuProps) => {
  const { isNotLargeScreen } = useHeaderContext();
  const { setNavigationContent } = useSetHeaderContext();

  useEffect(() => {
    const navigationContent = getChildElementsEvenIfContainersInbetween(children);
    setNavigationContent(navigationContent);
  }, [children]);

  if (isNotLargeScreen) return null;

  return (
    <div className={styles.headerNavigationMenuContainer}>
      <nav role="navigation" aria-label={ariaLabel} id={id} className={styles.headerNavigationMenu}>
        <ul className={styles.headerNavigationMenuList}>
          <HeaderNavigationMenuContextProvider>
            <HeaderNavigationMenuContent />
          </HeaderNavigationMenuContextProvider>
        </ul>
      </nav>
    </div>
  );
};
