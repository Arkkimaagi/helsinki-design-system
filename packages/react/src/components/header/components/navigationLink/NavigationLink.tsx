import React, { MouseEventHandler, cloneElement, useCallback, useEffect, useRef, useState } from 'react';
// import base styles
import '../../../../styles/base.css';
import { v4 as uuidv4 } from 'uuid';

// import core base styles
import 'hds-core';
import styles from './NavigationLink.module.scss';
import { styleBoundClassNames } from '../../../../utils/classNames';
import { Link } from '../../../link';
import { NavigationLinkDropdown, NavigationLinkInteraction, DropdownMenuPosition } from './navigationLinkDropdown';
import { useHeaderNavigationMenuContext } from '../headerNavigationMenu/HeaderNavigationMenuContext';
import { useHeaderContext } from '../../HeaderContext';
import { MergeElementProps } from '../../../../common/types';

const classNames = styleBoundClassNames(styles);

export type LinkProps = {
  /**
   * Indicator for active link. This is used in HeaderNavigationMenu.
   */
  active?: boolean;
  /**
   * Additional class names to apply.
   */
  className?: string;
  /**
   * Boolean for indicating whether the link has a dropdown menu.
   */
  hasDropdownLinks?: boolean;
  /**
   * Hypertext Reference of the link.
   */
  href: string;
  /**
   * Boolean for indicating whether the link's dropdown menu is open.
   */
  isDropdownOpen?: boolean;
  /**
   * Label for link.
   */
  label: string;
  /**
   * Optional event handler for onMouseEnter.
   */
  onMouseEnter?: MouseEventHandler;
  /**
   * Depth in nested dropdowns.
   * @internal
   */
  depth: number;
};

export type NavigationLinkProps<ReactElement> = {
  /**
   * Indicator for active link. This is used in HeaderNavigationMenu.
   */
  active?: boolean;
  /**
   * Element type to use instead of the default HDS Link.
   * @default Link
   * @example
   * ```ts
   * as={CustomLink}
   * ```
   */
  as?: ReactElement;
  /**
   * Additional class names to apply for the link element.
   */
  className?: string;
  /**
   * Aria-label for the dropdown button to describe closing the dropdown.
   */
  closeDropdownAriaButtonLabel?: string;
  /**
   * Depth in nested dropdowns.
   * @internal
   */
  depth?: number;
  /**
   * Additional class name for the dropdown element.
   */
  dropdownClassName?: string;
  /**
   * Additional class name for the dropdown items.
   */
  dropdownLinkClassName?: string;
  /**
   * Array of NavigationLink components to render in a dropdown. Can be used only inside navigation components.
   */
  dropdownLinks?: Array<React.ReactElement>;
  /**
   * Hypertext Reference of the link.
   * @default #
   */
  href?: string;
  /**
   * Element index given by parent mapping.
   * @internal
   */
  index?: number;
  /**
   * Label for link.
   */
  label: string;
  /**
   * Aria-label for the dropdown button to describe opening the dropdown.
   */
  openDropdownAriaButtonLabel?: string;
  /**
   * Which sub navigation index is open.
   * @internal
   */
  openSubNavIndex?: number;
  /**
   * Set which sub navigation index is open.
   * @internal
   */
  setOpenSubNavIndex?: (val: number) => void;
  /**
   * Additional class name for the dropdown wrapper element.
   */
  wrapperClassName?: string;
};

export type HeaderNavigationLinkProps<ReactElement extends React.ElementType = 'a'> = MergeElementProps<
  ReactElement,
  NavigationLinkProps<ReactElement>
>;

export const NavigationLink = <T extends React.ElementType = 'a'>({
  active,
  as: LinkComponent,
  className,
  wrapperClassName,
  dropdownClassName,
  dropdownLinks,
  dropdownLinkClassName,
  href,
  index,
  label,
  openSubNavIndex,
  setOpenSubNavIndex,
  depth = 0,
  openDropdownAriaButtonLabel,
  closeDropdownAriaButtonLabel,
  ...rest
}: HeaderNavigationLinkProps<T>) => {
  const Item = React.isValidElement(LinkComponent) ? LinkComponent.type : LinkComponent;
  const { isNotLargeScreen } = useHeaderContext();
  const [isDropdownOpen, setDropdownOpen] = useState<boolean>(false);
  const [dropdownOpenedBy, setDropdownOpenedBy] = useState<null | NavigationLinkInteraction>(null);
  const [dynamicPosition, setDynamicPosition] = useState<null | DropdownMenuPosition>(null);
  const { openMainNavIndex, setOpenMainNavIndex } = useHeaderNavigationMenuContext();
  const containerRef = useRef<HTMLSpanElement>(null);
  const isSubNavLink = openSubNavIndex !== undefined && setOpenSubNavIndex !== undefined;

  const handleDynamicMenuPosition = (val: boolean) => {
    // Null position when false so if user resizes browser, the calculation is done again
    if (!val) setDynamicPosition(null);
    // In case of SSR just set the menu to the right
    else if (val && window === undefined) setDynamicPosition(DropdownMenuPosition.Right);
    else {
      // eslint-disable-next-line no-lonely-if
      if (containerRef.current) {
        // Calculate which side has more space for the menu
        const { x: leftPosition, width } = containerRef.current.getBoundingClientRect();
        const rightPosition = leftPosition + width;
        const position =
          leftPosition > window.innerWidth - rightPosition ? DropdownMenuPosition.Left : DropdownMenuPosition.Right;
        setDynamicPosition(position);
      }
    }
  };

  // Handle dropdown open state by calling either internal state or context
  const handleDropdownOpen = (val: boolean, interaction?: NavigationLinkInteraction) => {
    // Set menu position on nested dropdowns to right or left depending on screen size
    if (depth >= 1) handleDynamicMenuPosition(val);
    setDropdownOpenedBy(!val ? null : interaction);
    setDropdownOpen(val);
    // If sub navigation props given, call them
    if (isSubNavLink && index !== undefined) {
      setOpenSubNavIndex(val ? index : -1);
    }
    // Otherwise it's safe to assume that this link is from main navigation and we can call context
    else {
      // eslint-disable-next-line no-lonely-if
      if (((val !== isDropdownOpen && openMainNavIndex === index) || val) && setOpenMainNavIndex) {
        // If closing dropdown, call context only if this is the open main nav dropdown. No need for checks if opening though.
        setOpenMainNavIndex(val ? index : -1);
      }
    }
  };

  const closeDropdown = () => {
    setDropdownOpen(false);
    setDropdownOpenedBy(null);
  };

  useEffect(() => {
    // If sub navigation index is not provided, we need to react to main nav context changes.
    if (!isSubNavLink) {
      // Since only one navigation link menu should be open, close this one if it's not the one that's open.
      if (openMainNavIndex !== index && isDropdownOpen) {
        closeDropdown();
      }
    }
  }, [openMainNavIndex]);

  useEffect(() => {
    // If sub nav and it's index differs from this, this link's dropdown should close
    if (isSubNavLink && openSubNavIndex !== index && isDropdownOpen) {
      closeDropdown();
    }
  }, [openSubNavIndex]);

  // Close dropdown menu when clicked outside
  const handleOutsideClick = useCallback(
    (e: MouseEvent) => {
      if (isDropdownOpen && containerRef.current && !containerRef.current.contains(e.target as Node)) {
        closeDropdown();
      }
    },
    [isDropdownOpen],
  );

  /* Set event listener only when dropdown open */
  useEffect(() => {
    if (isDropdownOpen) document.addEventListener('click', handleOutsideClick);
    return () => document.removeEventListener('click', handleOutsideClick);
  }, [isDropdownOpen]);

  const navigationWrapperLinkClassName = classNames(
    { isNotLargeScreen },
    styles.navigationLinkWrapper,
    styles[`depth-${depth}`],
    wrapperClassName,
  );
  const navigationLinkClassName = classNames(styles.navigationLink, styles[`depth-${depth}`], className, {
    active,
    isNotLargeScreen,
  });

  return (
    <span
      className={navigationWrapperLinkClassName}
      {...(dropdownLinks &&
        dropdownOpenedBy === NavigationLinkInteraction.Hover && {
          onMouseLeave: () => handleDropdownOpen(false),
        })}
      ref={containerRef}
      {...(dropdownLinks &&
        dropdownOpenedBy !== NavigationLinkInteraction.Click && {
          onMouseEnter: () => handleDropdownOpen(true, NavigationLinkInteraction.Hover),
        })}
    >
      <Item
        className={navigationLinkClassName}
        href={href}
        {...(active && { 'aria-selected': true })}
        {...(Boolean(dropdownLinks) && { 'aria-expanded': isDropdownOpen })}
        {...rest}
      >
        {label}
      </Item>
      {dropdownLinks && (
        <NavigationLinkDropdown
          open={isDropdownOpen}
          setOpen={handleDropdownOpen}
          index={index}
          depth={depth + 1}
          className={dropdownClassName}
          dynamicPosition={dynamicPosition}
          openDropdownAriaButtonLabel={openDropdownAriaButtonLabel}
          closeDropdownAriaButtonLabel={closeDropdownAriaButtonLabel}
        >
          {dropdownLinks.map((child) => {
            return cloneElement(child as React.ReactElement, {
              key: uuidv4(),
              wrapperClassName,
              dropdownClassName,
              dropdownLinkClassName,
            });
          })}
        </NavigationLinkDropdown>
      )}
    </span>
  );
};

NavigationLink.defaultProps = {
  as: Link,
  href: '#',
};
