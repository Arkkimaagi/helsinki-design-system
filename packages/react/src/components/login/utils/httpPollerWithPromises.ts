import to from 'await-to-js';

import createHttpPoller, { HttpPoller, isSuccessfulHttpResponse } from './httpPoller';
import { isAbortError } from './abortFetch';

export type RetryingPollerProps = {
  pollFunction: () => Promise<Response | undefined>;
  pollIntervalInMs?: number;
  maxRetries?: number;
};

export default async function retryPollingUntilSuccessful({
  pollFunction,
  pollIntervalInMs = 1000,
  maxRetries = 10,
}: RetryingPollerProps): Promise<Response> {
  const [err, response] = await to(pollFunction());
  if (!err && response && isSuccessfulHttpResponse(response)) {
    return Promise.resolve(response);
  }
  if (!maxRetries || (err && isAbortError(err))) {
    return Promise.reject(err);
  }
  const abortStatusCode = -1;
  let retries = maxRetries;
  let poller: HttpPoller | undefined;
  const removePoller = () => {
    if (poller) {
      poller.stop();
    }
    poller = undefined;
  };
  return new Promise((resolve, reject) => {
    poller = createHttpPoller({
      pollFunction: async () => pollFunction(),
      onSuccess: (successResponse) => {
        retries = 0;
        resolve(successResponse);
        return { keepPolling: false };
      },
      onError: (status, error) => {
        if (status === abortStatusCode) {
          retries = 0;
          reject(error);
          return { keepPolling: false };
        }
        retries -= 1;
        const keepPolling = retries > 0;
        if (!keepPolling) {
          reject(new Error('Max retries reached'));
        }
        return { keepPolling };
      },
      shouldPoll: () => retries > 0,
      pollIntervalInMs,
      onErrorStatusWhenAborted: abortStatusCode,
    });
    poller.start();
  })
    .then((r) => Promise.resolve(r as Response))
    .catch((e) => Promise.reject(e))
    .finally(() => removePoller());
}
