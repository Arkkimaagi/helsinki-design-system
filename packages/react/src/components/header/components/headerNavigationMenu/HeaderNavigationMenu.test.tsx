import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import { HeaderNavigationMenu } from './HeaderNavigationMenu';
import { NavigationLink } from '../navigationLink/NavigationLink';
import { HeaderNavigationMenuWrapper } from '../../../../utils/test-utils';

describe('<HeaderNavigationMenu /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<HeaderNavigationMenu />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(<HeaderNavigationMenu />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
  it('should open dropdown when the button is clicked', async () => {
    render(
      <NavigationLink
        href="#"
        label="Link"
        dropdownLinks={[
          <NavigationLink href="#" label="Test" dropdownLinks={[<NavigationLink href="#" label="Test" />]} />,
        ]}
      />,
      { wrapper: HeaderNavigationMenuWrapper },
    );
    userEvent.click(screen.getByTestId('dropdown-button-main-nav-0'));

    expect(screen.getByTestId('dropdown-menu-main-nav-0')).toBeVisible();
  });
  it('should have only one main dropdown active', async () => {
    render(
      <>
        <NavigationLink href="#" label="Link 1" dropdownLinks={[<NavigationLink href="#" label="Link 1 nested" />]} />
        <NavigationLink href="#" label="Link 2" dropdownLinks={[<NavigationLink href="#" label="Link 2 nested" />]} />
      </>,
      { wrapper: HeaderNavigationMenuWrapper },
    );
    // Click the first main navigation link's dropdown button
    userEvent.click(screen.getByTestId('dropdown-button-main-nav-0'));
    expect(screen.getByTestId('dropdown-menu-main-nav-0')).toBeVisible();
    // Click the second main navigation link's dropdown button
    userEvent.click(screen.getByTestId('dropdown-button-main-nav-1'));
    // Now the second dropdown should show
    expect(screen.getByTestId('dropdown-menu-main-nav-1')).toBeVisible();
    // The first dropdown should not be visible. This can only be tested if display: none; is set inline!
    expect(screen.getByTestId('dropdown-menu-main-nav-0')).not.toBeVisible();
  });
});
