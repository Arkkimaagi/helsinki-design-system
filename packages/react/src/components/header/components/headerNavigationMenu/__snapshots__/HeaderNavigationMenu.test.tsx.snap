// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderNavigationMenu /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="headerNavigationMenuContainer"
  >
    <nav
      class="headerNavigationMenu"
      role="navigation"
    >
      <ul
        class="headerNavigationMenuList"
      />
    </nav>
  </div>
</DocumentFragment>
`;
